### DISCUSSION TITLE
Make inclusive/exclusive ranges consistent and explicit

### DISCUSSION THREAD

**Discussion by zanebclark** [openmobilityfoundation/curb-data-specification#61] [Data Specification]
Range inclusivity/exclusivity is suggested in the [curbs](https://github.com/openmobilityfoundation/curb-data-specification/tree/main/curbs) spec. I'd recommend that range boundaries be explicitly defined as inclusive or exclusive. To aid in the sensible definition of contiguous ranges, I'd suggested that one range boundary is exclusive and one is inclusive. For example, defining two contiguous [Time Spans](https://github.com/openmobilityfoundation/curb-data-specification/tree/main/curbs#time-span) with inclusive `from` and inclusive `to` would be a bit cumbersome: 

```
{
  "from": "2021-08-11 00:00:00",
  "to": "2021-08-12 00:00:00"
}
```

```
{
  "from": "2021-08-12 00:00:01",
  "to": "2021-08-13 00:00:00"
}
```

Note the need to define the 2nd Time Span with a contrived `from`. Having an inclusive `from` and exclusive `to` would solve this. 

A few ranges with the my interpreted range boundary conditions: 
- [Curb Zone](https://github.com/openmobilityfoundation/curb-data-specification/tree/main/curbs#curb-zone)
    - `start_date`: Inclusive
    - `end_date`: Exclusive 
- [Time Span](https://github.com/openmobilityfoundation/curb-data-specification/tree/main/curbs#time-span)
    - `from`: Inclusive
    - `to`: Inclusive
- [Rate](https://github.com/openmobilityfoundation/curb-data-specification/tree/main/curbs#rate)
    - `start_minutes`: Inclusive
    - `end_minutes`: Inclusive
- [Location Reference](https://github.com/openmobilityfoundation/curb-data-specification/tree/main/curbs#location-reference)
    - `start`: Inclusive
    - `end`: Inclusive

**Comment by jiffyclub** [comment]
Big 👍  for explicitly defining inclusivity, and inclusive start/exclusive end is a good default.

In MDS we went through a similar discussion and ended up choosing to add flags that allow policy authors to choose whether start and end are inclusive or exclusive, defaulting to inclusive. There were cases where folks thought one situation or the other made for the simplest policy rules. Here's the PR where we made the change: https://github.com/openmobilityfoundation/mobility-data-specification/pull/689

**Reply by zanebclark** [reply]
I'm a big fan of that idea

**Reply by schnuerle** [reply]
I think for now we can just say start is inclusive and end is exclusive?

**Reply by schnuerle** [reply]
Check out [this PR](https://github.com/openmobilityfoundation/curb-data-specification/commit/42733c5ac8d19adc0f261545b4574d5e6fa76d04#diff-99381b09595a1c06da8cccb814a8f93a615f6b5f3cf29a8ab94a3f9d107f1edb) and the resulting [page starting here](https://github.com/openmobilityfoundation/curb-data-specification/tree/feature-release-work-1/curbs#curb-zone).

Made Curb Zone and Time Span inclusive and exclusive. Left Rate as inclusive all around (but maybe the end should be exclusive here too? eg at 60 minutes the billing stops, so it would be exclusive since 60:00, 60:01 ... 60:59 min:sec should not be billed?). Did not put it on Location Reference since it's measuring distance (and seems less urgent measuring centimeters).

Also references a new [Time Range](https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/general-information.md#time-range) section.

Any changes you have? Would like to make them today, but could be made in the next month as we go through release candidate approval.

**Reply by zanebclark** [reply]
@schnuerle , I'm so sorry for the delay! I was out of the office for a few days there. The PR looks good. A couple of thoughts on [Time Ranges](https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/general-information.md#time-ranges): 
- The TOC anchor doesn't work for me
- >'inclusive' at the start and 'exclusive' at the end, unless otherwise noted.

   I think this is the first part of @jiffyclub 's configurable inclusive/exclusive proposal: Define a default. Introducing a per-time-range inclusive/exclusive indicator field in the future wouldn't cause a breaking change, as undefined would be interpreted as the default. I like it.

**Reply by zanebclark** [reply]
>eg at 60 minutes the billing stops, so it would be exclusive since 60:00, 60:01 ... 60:59 min:sec should not be billed?)

Indeed. I think Rate ought to join the ranks of inclusive / exclusive ranges.

**Reply by schnuerle** [reply]
Thanks! Fixed TOC link and made rate end exclusive too (as I wrote my comment I realized it probably should be).

**Reply by jiffyclub** [reply]
Looks good, thanks @schnuerle! I think we will ultimately want to apply this to length as well since this issue comes up with any kind of continuous measurement. Does the span 10-20 cm end at the 19|20 boundary, or at the 20|21 boundary?

**Reply by schnuerle** [reply]
Ok you make a good point. Renamed 'Time Range' to 'Range Boundaries' in the General Info and Curbs areas of the spec, and included measurements as well. [See the latest](https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/curbs/README.md#curb-zone).

**Reply by schnuerle** [reply]
Added to [Metrics query parameters](https://github.com/openmobilityfoundation/curb-data-specification/blob/feature-release-work-1/metrics/README.md#query-parameters) too.

**Reply by zanebclark** [reply]
@schnuerle , I'm happy with the outcome here if you'd like to remove the feedback needed tag
